src/transactions/TransactionUtils.cpp

```diff
@@ -867,7 +867,18 @@ setAuthorized(LedgerTxnHeader const& header, LedgerTxnEntry& entry,
         throw std::runtime_error("trying to set invalid trust line flag");
     }
     auto& tl = entry.current().data.trustLine();
-    tl.flags = authorized;
+
+    const uint32_t authFlags =
+        AUTHORIZED_FLAG | AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG;
+
+    // clear the AUTHORIZED_* bits
+    uint32_t newFlag = tl.flags & (~authFlags);
+
+    // clear everything BUT the AUTHORIZED_* bits
+    authorized &= authFlags;
```

Comment:
At this point, we should have `authorized & authFlags == 0` anyway. Perhaps we should just throw at the top of this function if this ever happens? It would indicate a pretty serious error.

Reply:
If you do this, then this entire part of the function could be written as
```
tl.flags &= ~authFlags;
tl.flags |= authorized;
```
I think this would be a lot simpler. I spent a while trying to understand what was going on with `newFlag` before seeing how it all fit together.

Reply:
I agree, the code is confusing because it handles out of spec combinations.

As for contract, we should
* either disallow calling this function with an incompatible `authorized` value?
    * in that case, an explicit `if (newFlag != (authorized & ~authFlags)) { throw ... `
* or require `authorized` to only be AUTHORIZED flags (so require `(authorized & ~authFlags) == 0`)

Reply:
Added the throw and simplified according to your suggestions.